src/database.c

```diff
@@ -481,15 +521,185 @@ _add_to_db(ProfMessage* message, char* type, const Jid* const from_jid, const Ji
             sqlite3_free(err_msg);
         } else {
-            log_error("Unknown SQLite error");
+            log_error("Unknown SQLite error in _add_to_db().");
         }
     } else {
         int inserted_rows_count = sqlite3_changes(g_chatlog_database);
         if (inserted_rows_count < 1) {
             log_error("SQLite did not insert message (rows: %d, id: %s, content: %s)", inserted_rows_count, message->id, message->plain);
         }
+
+        sqlite3_int64 last_inserted_id = sqlite3_last_insert_rowid(g_chatlog_database);
+        if (last_inserted_id < 1) {
+            log_error("Message was inserted without error, but unable to get last inserted ID (got: %d)", last_inserted_id);
+            return;
+        }
+
+        if (message->replace_id) {
+            // TODO: maybe put as a trigger
+            auto_sqlite char* original_mark_query = sqlite3_mprintf("UPDATE `ChatLogs` SET `replaced_by_db_id`=%d WHERE `id` = %d",
+                                                                    last_inserted_id,
+                                                                    original_message_id);
+            if (!original_mark_query) {
+                log_error("Could not allocate memory for SQL replace old message query in log_database_add()");
+                return;
+            }
+            if (SQLITE_OK != sqlite3_exec(g_chatlog_database, original_mark_query, NULL, 0, &err_msg)) {
+                log_error("Unable to replace old message: %s", err_msg);
+                sqlite3_free(err_msg);
+                return;
+            }
+        }
+    }
+}
+
+static int
+_get_db_version(void)
+{
+    int current_version = -1;
+    const char* query = "SELECT `version` FROM `DbVersion` LIMIT 1";
+    sqlite3_stmt* statement;
+
+    if (sqlite3_prepare_v2(g_chatlog_database, query, -1, &statement, NULL) == SQLITE_OK) {
+        if (sqlite3_step(statement) == SQLITE_ROW) {
+            current_version = sqlite3_column_int(statement, 0);
+        }
+        sqlite3_finalize(statement);
+    }
+    return current_version;
+}
+
+/**
+ * Migration to version 2 introduces new field, indexes timestamps, to_jid, from_jid. Returns TRUE on success.
+ *
+ * Changed fields:
+ * `replaces_db_id` database ID for correcting message of the original message
+ * `replaced_by_db_id` database ID for original message of the last correcting message
+ */
+static gboolean
+_migrate_to_v2(void)
+{
+    char* err_msg;
+
+    // Make ChatLogsV2 table which will be renamed to ChatLogs on transformation success
+    char* query = "CREATE TABLE `ChatLogsV2` (`id` INTEGER PRIMARY KEY AUTOINCREMENT, `from_jid` TEXT NOT NULL, `to_jid` TEXT NOT NULL, `from_resource` TEXT, `to_resource` TEXT, `message` TEXT, `timestamp` TEXT, `type` TEXT, `stanza_id` TEXT, `archive_id` TEXT, `encryption` TEXT, `marked_read` INTEGER, `replaces_db_id` INTEGER, `replace_id` TEXT, `replaced_by_db_id` INTEGER)";
```

Comment:
I'd prefer if the name of the 'migration table' would be more clear. This looks like it'd be a legit table name ... maybe use `ChatLogs_migration_v1_to_v2`!?

Reply:
Done.